velox/dwio/parquet/writer/Writer.cpp

```diff
@@ -139,17 +139,131 @@ std::shared_ptr<WriterProperties> getArrowParquetWriterOptions(
   return properties->build();
 }
 
+void validateSchemaRecursive(const RowTypePtr& schema) {
+  // Check the schema's field names is not empty and unique.
+  VELOX_USER_CHECK_NOT_NULL(schema, "Field schema must not be empty.");
+  const auto& fieldNames = schema->names();
+
+  folly::F14FastSet<std::string> uniqueNames;
+  for (const auto& name : fieldNames) {
+    VELOX_USER_CHECK(!name.empty(), "Field name must not be empty.")
+    auto result = uniqueNames.insert(name);
+    VELOX_USER_CHECK(
+        result.second,
+        "File schema should not have duplicate columns: {}",
+        name);
+  }
+
+  for (auto i = 0; i < schema->size(); ++i) {
+    if (auto childSchema =
+            std::dynamic_pointer_cast<const RowType>(schema->childAt(i))) {
+      validateSchemaRecursive(childSchema);
+    }
+  }
+}
+
+std::vector<std::shared_ptr<::arrow::Field>> updateStructFieldName(
+    const std::vector<std::shared_ptr<::arrow::Field>>& fields,
+    const std::vector<std::string>& names) {
+  std::vector<std::shared_ptr<::arrow::Field>> newFields;
+  newFields.reserve(fields.size());
+  for (auto i = 0; i < fields.size(); ++i) {
+    newFields.push_back(::arrow::field(names[i], fields[i]->type()));
+  }
+
+  return newFields;
+}
+
+std::vector<std::shared_ptr<::arrow::Field>> updateComplexTypeFields(
+    const std::vector<std::shared_ptr<::arrow::Field>>& fields,
+    const RowTypePtr& rowType) {
+  std::vector<std::shared_ptr<::arrow::Field>> newFields;
+  newFields.reserve(rowType->size());
+  for (int i = 0; i < rowType->size(); ++i) {
+    auto field = fields[i];
+    if (auto childRowType =
+            std::dynamic_pointer_cast<const RowType>(rowType->childAt(i))) {
+      auto structType =
+          std::dynamic_pointer_cast<::arrow::StructType>(field->type());
+      auto newChildFields =
+          updateComplexTypeFields(structType->fields(), childRowType);
+      auto newStructFields =
+          updateStructFieldName(newChildFields, childRowType->names());
+      newFields.push_back(field->WithType(::arrow::struct_(newStructFields)));
+    } else if (
+        auto childArray =
+            std::dynamic_pointer_cast<const ArrayType>(rowType->childAt(i))) {
+      auto listType =
+          std::dynamic_pointer_cast<::arrow::BaseListType>(field->type());
+      auto valueField = field;
+      if (auto nestedRowType = std::dynamic_pointer_cast<const RowType>(
+              childArray->elementType())) {
```

Comment:
Does this work for ARRAY(ARRAY(ARRAY(ROW(...))))?

Reply:
@mbasmanova It seems not. I have updated to update the field name recursively. Can you help to review again?